3D Tiles - Add allVisibleTilesLoaded event to Cesium3DTileset

lilleyse opened the pull request:
For #4400
From: https://github.com/AnalyticalGraphicsInc/cesium/pull/4401#issuecomment-251687673

The `loadedProgress` event doesn't take attempted requests into account, like when a request is made but the request scheduler is full, so checking when a tileset is loaded by checking that number processing and number of active requests are 0 may be faulty.

This `allVisibleTilesLoaded` event accounts for attempted requests and active requests / processing.

@mramato let me know how this works.

pjcozzi commented:
Code looks OK at a glance.  Why not rename to `allTilesLoaded`?  "Visible" is likely implicit to the user.

@mramato please test and merge.

mramato commented:
I like the idea of `viewComplete` since we can re-use that name for imagery, terrain, and anywhere else people want to know if the scene is complete or not.  Also, I would prefer a boolean to an event, since my use case is to check in the postRender loop instead of having to subscribe to every tileset I create.

I'm checking out the behavior now.

mramato commented:
Other than the desire for checkable flag, this resolves the original problem I was seeing.  Thanks @lilleyse!

lilleyse commented:
I added a `viewComplete` flag to `Cesium3DTileset`. In order to make it work correctly on the first frame I changed the load process slightly so that the tileset isn't considered ready before its first update. Now the load state and ready look more like `Model`.

mramato commented:
Given your latest changes, do you need @pjcozzi to look at this again before I test and merge?

lilleyse commented:
Yes that's a good idea. Feel free to test though.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -472,6 +479,23 @@ define([
+         *
+         * @default false
+         */
+        viewComplete : {
```
Can we please name this differently since it mixes poorly with the camera view semantics.  Why not stay consistent with the well understood tiles loaded paradigm?

pjcozzi replied:
For sure.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -472,6 +479,23 @@ define([
         },
 
         /**
+         * When <code>true</code>, all tiles in view are are loaded.
```
Same comment.

Probably have a `@see` for each as well.

pjcozzi replied:
> Probably have a @see for each as well.

Unless they are right next to each other in the doc like `ready` and `readyPromise`.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -1468,8 +1527,23 @@ define([
      * @exception {DeveloperError} The tileset must be 3D Tiles version 0.0.  See https://github.com/AnalyticalGraphicsInc/3d-tiles#spec-status
      */
     Cesium3DTileset.prototype.update = function(frameState) {
+        if (!this._ready) {
```
Is all this really needed?  Couldn't we just have `_ready` and move some logic into `frameState.afterRender` to preserve the logic of the first `update`?

Either way, there should be a comment so we don't try to remove the "needless" complexity later.

lilleyse replied:
Yeah looking at it now, it's more complicated than it needs to be. It's updated now.

pjcozzi replied:
Looks nice.

lilleyse commented:
Updated with the name change.

pjcozzi commented:
Thanks for the updates, @lilleyse!

@mramato can you retest with your scenario and merge when ready?

pjcozzi commented:
@mramato any update?

mramato commented:
Sorry, will get to this soon, tomorrow at the latest.

mramato commented:
Just that one comment, this solves the original problem I had.  Thanks @lilleyse!

mramato commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -335,32 +355,32 @@ define([
             that._properties = tilesetJson.properties;
             that._geometricError = tilesetJson.geometricError;
             that._root = data.root;
-            that._readyPromise.resolve(that);
```
@lilleyse any update here?  If resolving inside of the update is important, another solution would be to make sure we only reject inside the update as well, so it's at least consistent.  But I still think having to add it to the scene to have the promise resolve is odd (if not simply inconsistent with imagery)

mramato replied:
@lilleyse bump

lilleyse replied:
I haven't forgot about this yet - I'll try to get to it soon.

pjcozzi commented:
@lilleyse is it trivial to finish and merge this?

lilleyse commented:
Updated.

@mramato The ready promise behavior should be back to normal now.

There is one potential side effect - `tilesVisible` will be true between when the ready promise resolves and the first update occurs. Will this be a problem? It can be solved by adding a variable that marks when update is called for the first time, but I didn't want to add that clutter unless needed.

At least, the `allTilesLoaded` event should work as expected and only fire after the first update.

lilleyse commented:
> There is one potential side effect - tilesVisible will be true between when the ready promise resolves and the first update occurs. Will this be a problem? It can be solved by adding a variable that marks when update is called for the first time, but I didn't want to add that clutter unless needed.

Fixes this and made the tests better at checking these corner cases. Ready now.

pjcozzi commented:
LGTM
